Inferred json reader crashes on multiple nested fields

I noticed this while working on #714.

```
$ echo '{"a.b.c": 1, "a.b.d": 2}' | zq -t -
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
github.com/brimsec/zq/zio/ndjsonio.(*inferParser).parseObject(0xc000032608, 0xc000342000, 0x18, 0x10000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x3, ...)
        /Users/henridf/work/brim/zq/zio/ndjsonio/inferparser.go:62 +0xe2c
github.com/brimsec/zq/zio/ndjsonio.(*Reader).Parse(0xc000032600, 0xc000342000, 0x18, 0x10000, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00003c090, ...)
        /Users/henridf/work/brim/zq/zio/ndjsonio/reader.go:112 +0x184
github.com/brimsec/zq/zio/ndjsonio.(*Reader).Read(0xc000032600, 0x13d9f42, 0x18f8ce0, 0xc00003c090)
        /Users/henridf/work/brim/zq/zio/ndjsonio/reader.go:126 +0xac
github.com/brimsec/zq/zio/detector.match(0x1b903c0, 0xc000032600, 0x195f8a4, 0x6, 0x0, 0x0)
        /Users/henridf/work/brim/zq/zio/detector/reader.go:70 +0x38
github.com/brimsec/zq/zio/detector.NewReader(0x1b90340, 0xc0002fb200, 0xc0002e2540, 0xc0002fb200, 0xc0000be000, 0x0, 0x1)
        /Users/henridf/work/brim/zq/zio/detector/reader.go:49 +0x588
github.com/brimsec/zq/zio/detector.OpenFromNamedReadCloser(0xc0002e2540, 0x1b99b20, 0xc0000be000, 0x7ffeefbffa86, 0x1, 0x195d3d9, 0x4, 0x1, 0x0, 0x1998ef3, ...)
        /Users/henridf/work/brim/zq/zio/detector/file.go:131 +0xaf
github.com/brimsec/zq/zio/detector.OpenFile(0xc0002e2540, 0x7ffeefbffa86, 0x1, 0x195d3d9, 0x4, 0x1, 0x0, 0x1998ef3, 0x59, 0x0, ...)
        /Users/henridf/work/brim/zq/zio/detector/file.go:68 +0x11a
main.(*Command).inputReaders(0xc0002e40a0, 0xc0000ce050, 0x1, 0x1, 0x0, 0x0, 0xaa00000001927a60, 0x1115f01, 0xaa7e7729fbbf34ca)
        /Users/henridf/work/brim/zq/cmd/zq/zq.go:270 +0x166
main.(*Command).Run(0xc0002e40a0, 0xc0000ce050, 0x1, 0x1, 0x0, 0x0)
        /Users/henridf/work/brim/zq/cmd/zq/zq.go:218 +0x2cf
github.com/mccanne/charm.(*instance).run(0xc0000ede20, 0xc0000ce040, 0x2, 0x2, 0x0, 0x0)
        /Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/instance.go:53 +0x28c
github.com/mccanne/charm.(*Spec).ExecRoot(0x2155000, 0xc0000ce040, 0x2, 0x2, 0xc0002cdf50, 0x100742f, 0xc0000a6058, 0x0)
        /Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/charm.go:77 +0x96
main.main()
        /Users/henridf/work/brim/zq/cmd/zq/main.go:9 +0x78
```

--- a/zio/ndjsonio/inferparser.go
+++ b/zio/ndjsonio/inferparser.go
@@ -6,8 +6,8 @@ import (
 	"sort"
 
 	"github.com/brimsec/zq/pkg/byteconv"
+	"github.com/brimsec/zq/proc"
 	"github.com/brimsec/zq/zcode"
-	"github.com/brimsec/zq/zio/zeekio"
 	"github.com/brimsec/zq/zng"
 	"github.com/brimsec/zq/zng/resolver"
 	"github.com/buger/jsonparser"
@@ -36,56 +36,34 @@ func (p *inferParser) parseObject(b []byte) (zng.Value, error) {
 	sort.Slice(kvs, func(i, j int) bool {
 		return bytes.Compare(kvs[i].key, kvs[j].key) < 0
 	})
-
-	// Build the list of columns (without types yet) and then run them
-	// through Unflatten() to find nested records.
-	columns := make([]zng.Column, len(kvs))
-	for i, kv := range kvs {
-		columns[i] = zng.NewColumn(string(kv.key), zng.TypeString)
-	}
-	columns, _, err = zeekio.Unflatten(p.zctx, columns, false)
-	if err != nil {
-		return zng.Value{}, err
-	}
-
-	// Parse the actual values and fill in column types along the way,
-	// taking care to step into nested records as necessary.
-	colno := 0
-	nestedColno := 0
-	var vals, nestedVals []zng.Value
+	var fields []string
+	var zngTypes []zng.Type
+	var zngValues []zng.Value
 	for _, kv := range kvs {
-		val, err := p.parseValue(kv.value, kv.typ)
+		fields = append(fields, string(kv.key))
+		v, err := p.parseValue(kv.value, kv.typ)
 		if err != nil {
 			return zng.Value{}, err
 		}
-
-		recType, isRecord := columns[colno].Type.(*zng.TypeRecord)
-		if isRecord {
-			nestedVals = append(nestedVals, val)
-		} else {
-			vals = append(vals, val)
-		}
-
-		if isRecord {
-			recType.Columns[nestedColno].Type = val.Type
-			nestedColno += 1
-			if nestedColno == len(recType.Columns) {
-				vals = append(vals, zng.Value{recType, encodeContainer(nestedVals)})
-				nestedVals = []zng.Value{}
-				nestedColno = 0
-				colno += 1
-			}
-		} else {
-			columns[colno].Type = val.Type
-			colno += 1
-		}
+		zngTypes = append(zngTypes, v.Type)
+		zngValues = append(zngValues, v)
 	}
-
-	typ, err := p.zctx.LookupTypeRecord(columns)
+	columnBuilder, err := proc.NewColumnBuilder(p.zctx, fields)
 	if err != nil {
 		return zng.Value{}, err
 	}
-	return zng.Value{typ, encodeContainer(vals)}, nil
+	typ, err := p.zctx.LookupTypeRecord(columnBuilder.TypedColumns(zngTypes))
+	if err != nil {
+		return zng.Value{}, err
+	}
+	for _, v := range zngValues {
+		columnBuilder.Append(v.Bytes, zng.IsContainerType(zng.AliasedType(v.Type)))
+	}
+	bytes_, err := columnBuilder.Encode()
+	if err != nil {
+		return zng.Value{}, err
+	}
+	return zng.Value{Type: typ, Bytes: bytes_}, nil
 }
 
 func (p *inferParser) parseValue(raw []byte, typ jsonparser.ValueType) (zng.Value, error) {

--- a/zio/ndjsonio/inferparser.go
+++ b/zio/ndjsonio/inferparser.go
@@ -59,11 +59,11 @@ func (p *inferParser) parseObject(b []byte) (zng.Value, error) {
 	for _, v := range zngValues {
 		columnBuilder.Append(v.Bytes, zng.IsContainerType(zng.AliasedType(v.Type)))
 	}
-	bytes_, err := columnBuilder.Encode()
+	zbytes, err := columnBuilder.Encode()
 	if err != nil {
 		return zng.Value{}, err
 	}
-	return zng.Value{Type: typ, Bytes: bytes_}, nil
+	return zng.Value{Type: typ, Bytes: zbytes}, nil
 }
 
 func (p *inferParser) parseValue(raw []byte, typ jsonparser.ValueType) (zng.Value, error) {

--- a/zio/ndjsonio/ndjson_test.go
+++ b/zio/ndjsonio/ndjson_test.go
@@ -99,6 +99,11 @@ func TestNDJSON(t *testing.T) {
 			output: `{ "s": "foo", "nest": { "s": "bar", "n": 5 }}`,
 		},
 		{
+			name:   "legacy nested fields with multiple levels of nesting",
+			input:  `{ "a.b.1": 1, "a.b.2": 2, "a.b.c.3": 3, "a.b.c.4": 4 }`,
+			output: `{ "a": { "b": { "1": 1, "2": 2, "c": { "3": 3, "4": 4 } } } }`,
+		},
+		{
 			name:   "string with unicode escape",
 			input:  `{ "s": "Hello\u002c world!" }`,
 			output: `{ "s": "Hello, world!" }`,

